src/Bicep.Core/TypeSystem/UnassignableType.cs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Bicep.Core.Diagnostics;
+
+namespace Bicep.Core.TypeSystem
+{
+    public class UnassignableTypeSymbol : TypeSymbol
+    {
+        private const string ErrorTypeName = "error";
+        private readonly ImmutableArray<ErrorDiagnostic> errors;
+        private readonly TypeKind kind;
```

Comment:
I don't think you need to create this extra field. You should be able to replace:
```
public override TypeKind TypeKind => kind
```
with
```
public override TypeKind TypeKind { get; }
```
And set it directly in the constructor with:
```
this.TypeKind = kind;
```